### ISSUE TITLE
[RFC] How to deal with outliers when plotting convergence curves

### ISSUE THREAD

**Issue by jeandut** [benchopt/benchopt#539]
Sometimes for some hyper-parameters solvers diverge (i.e. too large learning rates).
Even when catching the divergence in custom StoppingCriterion logic and stopping the solver, the first value might still be too large with respect to the rest of the points leading to sub-optimal choice of axis limits when a new user looks at the curves.
I give a real-life example below:  

**Properly-sized Benchopt plot:**
<img width="1425" alt="Capture d’écran 2023-02-21 à 16 27 02" src="https://user-images.githubusercontent.com/11030901/220387104-47b76dc1-e138-4eb1-8be6-f47c7f5e5229.png">
**Adding back a diverging solver:**
<img width="1411" alt="Capture d’écran 2023-02-21 à 16 28 16" src="https://user-images.githubusercontent.com/11030901/220387415-665c6a7f-8883-4650-a04c-bc2baa7132f2.png">
One can always zoom in using the interface but it is a bit disappointing that one should have to zoom.

I see two ways to circumvent this issue:
1. Allow the user to set default y-limits: `benchopt run --ylim 0.2 0.5` for instance
2. Some kind of outlier detection in the spirit of [this answer](https://stackoverflow.com/questions/11882393/matplotlib-disregard-outliers-when-plotting) with potential Warnings in the plot or arrows pointing towards non displayed outliers.

Thanks a lot for the great tool !

**Comment by jeandut** [comment]
Note that 1. would only work for `objective_value` and not for other metrics potentially.

**Comment by mathurinm** [comment]
Passing options via command line may clutter the CLI, maybe we could have plot options passed through a `plot_config.yml` file ?

**Comment by tomMoral** [comment]
There is a embryo-code for plot configuration on a per-benchmark basis:
https://github.com/benchopt/benchopt/blob/1487639e156a9acec809cb3a86fd2d2e7960115f/benchopt/config.py#L43-L56

The original goal was to select the default plot to display for the benchmark so it is probably easy to repurpose this to configure the default plot.

One thing that is not so easy is how to define bad outliers from good ones. What happens if one solver is very fast while the others are slow? I think this kind of magic trick of filtering should be possible easily, but not by default. WDYT?

Another question is what should be configurable? In my list right now:
- default plot (type, objective_column, dataset, parameters)
- default scale (linear, semilogN, loglog)
- default axis limits (x/ylim)
Do you see other stuff?

Finally, could you share the CSV that produces the HTML above? this will help debug the functionality :)

**Comment by jeandut** [comment]
Yes outlier detection is a research domain in itself I agree and it depends heavily on what the user wants. 
It would be much simpler from an implementation perspective to allow the user to give default config parameters **however** this has the limitation that it would apply to only one metric aka the first one (or the one indicated in the default config file) and only to its absolute value (not distance to optimum etc.), one user might log metrics that are on different scales...

In my very narrow view of things I would like some kind of simple mechanism for outlier detection based on quantile of the y values (and maybe for time as well ? but if operating on the x then it might create edge case or whatnot).  

Since it would only apply to a view and obviously not to the parquet itself, the only bad things I see is if it hides some information from the user aka the user thinks there is no outlier. Therefore if outlier detection is performed then one should display clearly that some outliers are not visible and that the user can click on a button to make the autoscale not remove outliers. Here a toggle might be a good option with a default parametrizable by the user ; )

Regarding what the user could pass I agree with your list. Also filters for curves displayed would be nice (regexp or hyperparameters based). Ofc one could envision customizing labels or axis in the very very long-term ala matplotlib. 


Here is a csv file to reproduce the effect:
[outliers.csv](https://github.com/benchopt/benchopt/files/10822116/outliers.csv)
